sklearn/metrics/pairwise.py

```diff
@@ -917,20 +932,27 @@ def gower_distances(X, Y=None, categorical_features=None, scale=True):
         if X.shape[0] != Y.shape[0] or X is not Y:
             j_start = 0
 
-        # Calculates the similarities for categorical columns
-        cat_dists = (X[i, cat_mask] != Y[j_start:, cat_mask]) | \
-                    (_object_dtype_isnan(X[i, cat_mask]) |
-                     _object_dtype_isnan(Y[j_start:, cat_mask]))
+        # Makes the comparisson for np.nan for arrays with dtype=np.object,
+        # this is necessary as some deployments returns True for
+        # np.nan == np.nan
+        cat_nan_cols = (_object_dtype_isnan(X[i, cat_mask]) |
+                        _object_dtype_isnan(Y[j_start:, cat_mask]))
 
+        # Calculates the similarities for categorical columns
+        cat_dists = ((X[i, cat_mask] != Y[j_start:, cat_mask]) | cat_nan_cols)
         # Calculates the Manhattan distances for numerical columns
-        num_dists = abs(X[i, num_mask].astype(np.float32) -
-                        Y[j_start:, num_mask].astype(np.float32)) / ranges
+        num_dists = abs(X[i, num_mask] -
+                        Y[j_start:, num_mask]) / ranges
+
         # Calculates the number of non missing columns
-        n_missing = X.shape[1] - (np.isnan(cat_dists).sum(axis=1) +
-                                  np.isnan(num_dists).sum(axis=1))
+        n_missing = X.shape[1] - (cat_nan_cols.sum(axis=1) +
+                                  _object_dtype_isnan(num_dists).sum(axis=1))
         # Gets the final results
-        results = np.sum(cat_dists, axis=1) + np.sum(num_dists, axis=1)
-        results /= n_missing
+        sum = np.sum(cat_dists, axis=1) + np.sum(num_dists, axis=1)
+
+        results = np.divide(sum, n_missing,
```

Comment:
I'm finding this very hard to read.

Reply:
Do you just mean to do something like

```py
sum[n_missing == 0] = np.nan
sum /= n_missing
```
such that `results` is no longer needed?

Reply:
https://github.com/scikit-learn/scikit-learn/pull/9555#discussion_r316609898
All the arrays are np.object now, when numerical columns are present, because I removed the float32 conversion. Because of this, if I use a / operand I'll get a divide by zero error, if the number of cols are zero. 

Previously with the numerical arrays as float32 there won't be a divide by zero error, and the result of division will be np.nan , that is not the case anymore with the arrays as np.object.

Reply:
If I use the proposed solution I still get ZeroDivisionError:

```
            sum[non_missing == 0] = np.nan
>           sum /= non_missing
E           ZeroDivisionError: float division by zero

```

Reply:
If I simply make the division straightaway I also will get ZeroDivisionError:
```
            print(sum.dtype)
            print(non_missing.dtype)
>           results = sum / non_missing
E           ZeroDivisionError: float division by zero

sklearn\metrics\pairwise.py:955: ZeroDivisionError
---------------------------------------------------------------- Captured stdout call ----------------------------------------------------------------
object
int32
```

It can be noticed above that the dtype of  sum array is is an object, these  are the values that made  gower_distances function fail using the traditional / operator:

```
    X = [['M', False, 222.22, 1],
         ['F', True, 333.22, 2],
         ['M', True, 1934.0, 4],
         [np.nan, np.nan, np.nan, np.nan]]

```